File: app/src/main/java/cn/mijack/imagedrive/adapter/FirebaseStorageAdapter.java
package cn.mijack.imagedrive.adapter;

import android.app.Activity;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Query;

import cn.mijack.imagedrive.R;
import cn.mijack.imagedrive.entity.FirebaseImage;
import cn.mijack.imagedrive.ui.ImageDisplayActivity;

/**
 * @author Mr.Yuan
 * @date 2017/5/7
 */
public class FirebaseStorageAdapter extends FirebaseRecyclerAdapter<FirebaseImage, FirebaseStorageAdapter.StorageHolder> {
    public final static String TAG = "FirebaseStorageAdapter";

    public FirebaseStorageAdapter(Query ref) {
        super(FirebaseImage.class, R.layout.item_driver, StorageHolder.class, ref);
    }

    @Override
    protected void populateViewHolder(StorageHolder storageHolder, FirebaseImage firebaseImage, int position) {
        DatabaseReference reference = getRef(position);
        storageHolder.loadImage(firebaseImage,reference);
    }


    public static class StorageHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        private ImageView image;
        private FirebaseImage firebaseImage;
        private DatabaseReference reference;

        public StorageHolder(View itemView) {
            super(itemView);
            image = (ImageView) itemView.findViewById(R.id.image);
            image.setOnClickListener(this);
        }

        private static final String TAG = "StorageHolder";

        public void loadImage(FirebaseImage firebaseImage, DatabaseReference reference) {
            this.firebaseImage = firebaseImage;
            this.reference = reference;
            String url = firebaseImage.getDownloadUrl();
            Log.d(TAG, "loadImage: url" + url);
            Glide.with(itemView.getContext()).load(url).into(image);
        }

        @Override
        public void onClick(View v) {
            if (v.getId()==R.id.image){
                ActivityOptionsCompat activityOptions =ActivityOptionsCompat
                        .makeSceneTransitionAnimation((Activity)v.getContext(),v,"image");
                ImageDisplayActivity.showFirebaseImage(v.getContext(),firebaseImage,reference.toString(),activityOptions.toBundle());
            }
        }
    }
}

File: app/src/main/java/cn/mijack/imagedrive/adapter/FirebaseStorageAdapter.kt
package cn.mijack.imagedrive.adapter

import android.app.Activity
import android.support.v4.app.ActivityOptionsCompat
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.widget.ImageView
import cn.mijack.imagedrive.R
import cn.mijack.imagedrive.entity.FirebaseImage
import cn.mijack.imagedrive.ui.ImageDisplayActivity
import com.bumptech.glide.Glide
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.Query

/**
 * @author admin
 * @date 2017/8/30
 */

class FirebaseStorageAdapter(ref: Query) : FirebaseRecyclerAdapter<FirebaseImage, FirebaseStorageAdapter.StorageHolder>(FirebaseImage::class.java, R.layout.item_driver, StorageHolder::class.java, ref) {

    override fun populateViewHolder(storageHolder: StorageHolder, firebaseImage: FirebaseImage, position: Int) {
        val reference = getRef(position)
        storageHolder.loadImage(firebaseImage, reference)
    }


    class StorageHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val image: ImageView = itemView.findViewById<View>(R.id.image) as ImageView
        private lateinit var firebaseImage: FirebaseImage
        private var reference: DatabaseReference? = null

        init {
            image.setOnClickListener(this)
        }

        fun loadImage(firebaseImage: FirebaseImage, reference: DatabaseReference) {
            this.firebaseImage = firebaseImage
            this.reference = reference
            val url = firebaseImage.downloadUrl
            Log.d(TAG, "loadImage: url" + url)
            Glide.with(itemView.context).load(url).into(image)
        }

        override fun onClick(v: View) {
            if (v.id == R.id.image) {
                val activityOptions = ActivityOptionsCompat
                        .makeSceneTransitionAnimation(v.context as Activity, v, "image")
                ImageDisplayActivity.showFirebaseImage(v.context, firebaseImage, reference!!.toString(), activityOptions.toBundle())
            }
        }

        companion object {
            private val TAG = "StorageHolder"
        }
    }

    companion object {
        val TAG = "FirebaseStorageAdapter"
    }
}
